### PR TITLE
feat: Fall back to datasets' "extra" if arg extra is empty

### LINKED ISSUE #37006: DatasetManager.register_dataset_change(): Fall back to datasets' "extra" if arg extra is empty
### Description

I would like to know if it is possible to change the behavior of the `register_dataset_change` function to copy the `extra` from the specified dataset if the actual `extra` argument is empty?

### Use case/motivation

We have a lot of datasets where we, within the dataset, have versions of the data. 
Because - for lineage etc - we want to keep the URI for all of those the same (it is the same dataset after all), we use the `extra` argument to pass along the version number etc. Unfortunately, after a change is registered, we lose that information. 

It seems logical to us that when no `extra` is specified (which is *actually impossible* as far as we can see, since the current calling locations don't even allow it), that you fall back to the extra of the dataset. Our override is as simple as

```python
class DatasetManager(OriginalDatasetManager):
    def register_dataset_change(self, *, task_instance: TaskInstance, dataset: Dataset, extra=None, session: Session,
                                **kwargs) -> None:
        if extra is None and dataset.extra:
            extra = dataset.extra

        super().register_dataset_change(task_instance=task_instance, dataset=dataset, extra=extra, session=session,
                                        **kwargs)
```

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by zhenglin-charlie-li**
fix: https://github.com/apache/airflow/issues/37006

Fall back to datasets' "extra" if arg extra is empty

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by hussein-awala — COMMENTED** [review]
Could you add a test for this change?

**Comment by ZhengLin-Li** [comment]
@hussein-awala Thanks for your review. I have added some test cases.

File: airflow/datasets/manager.py
Comment (jedcunningham) [inline airflow/datasets/manager.py:1]: Personally, I don't think the dataset extra is a good fallback for a dataset event's extra. I see those extras as being different.

**Review by dirrao — CHANGES_REQUESTED** [review]
Looks like extras are reserved for future use. May be its worth to pass information about partition.

**Pull request closed without merging**